### PR TITLE
Add Datadog integration

### PR DESCRIPTION
**Pull request by tristanbes**
<p align="center">
  
</p>

The [Datadog](https://www.datadoghq.com/) integration was missing.

I added the support by adding a `DatadogHandler` and a `DatadogFormatter` that takes benefit of some of the Datadog features.

Here's what is outputed in the datadog log section: 

### Summary:
Log_Explorer___Datadog

### Log details: 
Log_Explorer___Datadog

### Exception:
Log_Explorer___Datadog

### CODE REVIEW

File: doc/02-handlers-formatters-processors.md [hunk 1]
```diff
@@ -154,6 +155,7 @@
 - [_FlowdockFormatter_](../src/Monolog/Formatter/FlowdockFormatter.php): Used to format log records into Flowdock messages, only useful for the FlowdockHandler.
 - [_MongoDBFormatter_](../src/Monolog/Formatter/MongoDBFormatter.php): Converts \DateTime instances to \MongoDate and objects recursively to arrays, only useful with the MongoDBHandler.
 - [_LogmaticFormatter_](../src/Monolog/Formatter/LogmaticFormatter.php): User to format log records to [Logmatic](http://logmatic.io/) messages, only useful for the LogmaticHandler.
```
Comment (Kocal) [inline doc/02-handlers-formatters-processors.md:157]:
```suggestion
- [_LogmaticFormatter_](../src/Monolog/Formatter/LogmaticFormatter.php): Used to format log records to [Logmatic](http://logmatic.io/) messages, only useful for the LogmaticHandler.
```

File: doc/02-handlers-formatters-processors.md [hunk 2]
```diff
@@ -154,6 +155,7 @@
...
+- [_Datadog_Formatter_](../src/Monolog/Formatter/DatadogFormatter.php): Used to format log records into [Datadog](https://www.datadoghq.com/) event json, with some [special](https://docs.datadoghq.com/logs/log_collection/?tab=http#how-to-get-the-most-of-your-application-logs) & [reserved](https://docs.datadoghq.com/logs/log_collection/?tab=http#reserved-attributes) attributes from Datadog.
```
Comment (Kocal) [inline doc/02-handlers-formatters-processors.md:158]:
```suggestion
- [_Datadog_Formatter_](../src/Monolog/Formatter/DatadogFormatter.php): Used to format log records into [Datadog](https://www.datadoghq.com/) JSON events, with some [special](https://docs.datadoghq.com/logs/log_collection/?tab=http#how-to-get-the-most-of-your-application-logs) & [reserved](https://docs.datadoghq.com/logs/log_collection/?tab=http#reserved-attributes) attributes from Datadog.
```

**Review by Kocal — COMMENTED** [review]
I think you need to add some tests, like other formatters/handlers.

File: doc/02-handlers-formatters-processors.md (hunk 1 above)
Comment (tristanbes) [reply doc/02-handlers-formatters-processors.md:157]: Unrelated to my PR.

**Comment by tristanbes** [comment]
Hello @Seldaek, is there anything you want me to add regarding this PR ?

File: src/Monolog/Formatter/DatadogFormatter.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * @var string|null
+     */
+    protected $loggerName;
+
+    /**
+     * @param string|null $applicationName The name of the application or service generating the log events used as the "source" Datadog attribute.
+     *                                     It is used to switch from Logs to APM, so make sure you define the same value when you use both products.
+     * @param string|null $systemName      The system/machine name, used as the "host" field of Datadog, defaults to the hostname of the machine
+     * @param string|null $env      The system/machine name, used as the "host" field of Datadog, defaults to the hostname of the machine
```
Comment (tristanbes) [inline src/Monolog/Formatter/DatadogFormatter.php:53]:
```suggestion
     * @param string|null $systemName      The system/machine name, used as the "host" field of Datadog, defaults to the hostname of the machine
     * @param string|null $env                     The environment name used as the "env" field of Datadog
```

File: src/Monolog/Formatter/DatadogFormatter.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+     * @var string|null
+     */
+    protected $loggerName;
+
+    /**
+     * @param string|null $applicationName The name of the application or service generating the log events used as the "source" Datadog attribute.
+     *                                     It is used to switch from Logs to APM, so make sure you define the same value when you use both products.
+     * @param string|null $systemName      The system/machine name, used as the "host" field of Datadog, defaults to the hostname of the machine
+     * @param string|null $env                     The environment name used as the "env" field of Datadog
+     * @param string|null $source          This corresponds to the integration name: the technology from which the log originated.
```
Comment (tristanbes) [inline src/Monolog/Formatter/DatadogFormatter.php:54]:
```suggestion
     * @param string|null $env               The environment name used as the "env" field of Datadog
     * @param string|null $source          This corresponds to the integration name: the technology from which the log originated.
```

File: src/Monolog/Formatter/DatadogFormatter.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * @var string|null
+     */
+    protected $loggerName;
+
+    /**
+     * @param string|null $applicationName The name of the application or service generating the log events used as the "source" Datadog attribute.
+     *                                     It is used to switch from Logs to APM, so make sure you define the same value when you use both products.
+     * @param string|null $systemName      The system/machine name, used as the "host" field of Datadog, defaults to the hostname of the machine
+     * @param string|null $env               The environment name used as the "env" field of Datadog
```
Comment (tristanbes) [inline src/Monolog/Formatter/DatadogFormatter.php:53]:
```suggestion
     * @param string|null $env             The environment name used as the "env" field of Datadog
```

Comment (tristanbes) [reply src/Monolog/Formatter/DatadogFormatter.php:53]: damn aligning docblock using web is a nightmare :D

**Comment by Seldaek** [comment]
Thanks for the PR but I am trying to avoid adding new handlers as it just adds maintenance burden and bloats the library.. So please rather submit this as your own package, then please feel free to add it to the wiki list at https://github.com/seldaek/monolog/wiki/Third-Party-Packages

That said, as I am a datadog user myself I was interested in the code and had a look. You can take this feedback or ignore it obviously, but IMO logging directly to datadog via tcp intake isn't the right approach. It slows down your application as TCP waits for packets to be received. It's faster (for the app layer) to log to filesystem and then configure the datadog agent to read the file and send it to datadog. Doing so also allows you to reuse the hostname and tags etc already configured on the machine level agent config.

Having a formatter that lets you output the logs in the best format for datadog consumption is a nice win though, I did this with a plain json log and a custom log rule on the datadog end but it's nicer to have the logs be correctly formatted.

**Comment by tristanbes** [comment]
Thank you for your feedback.

Unfortunatly, very few PaaS supports Datadog, that's why i had to create this formatter/handler

**Comment by alucic** [comment]
>  I did this with a plain json log and a custom log rule on the datadog end but it's nicer to have the logs be correctly formatted.

@Seldaek would you mind sharing your Datadog pipeline rules for JSONFormatter

**Comment by Seldaek** [comment]
I don't think I have any custom rules in places anymore. I do JsonFormatter to a StreamHandler to some text file on disk, which is logrotated. Then have in `/etc/datadog-agent/conf.d/customlogs.d/conf.yaml` something like this:

```
logs:
  - type: file
    path: /var/log/php_errors.log
    service: app
    source: php
    sourcecategory: php_errors
  - type: file
    path: /var/log/php_deprecations.log
    service: app
    source: php_deprecations
    sourcecategory: php_deprecations
```

I then have added columns to the log viewer options like `@extra.foo` which lets us display some extra data next to the message. The rest (stack trace and co) I usually just see by opening the log details. The message gets picked up correctly by default.

**Pull request closed without merging**